File: bin/mergeResults.py
#!/usr/bin/env python3

import argparse,re

def parseArguments():
    parser = argparse.ArgumentParser(description="Matches outputs from pepMatch and exact full Protein matches into a single gff file.")
    parser.add_argument("--pepMatchResults", required=True, help="Path to peptide match results")
    parser.add_argument("--fullSeqMatchResults", required=True, help="Path to full sequence match results")
    parser.add_argument("--peptideTab", required=True, help="Path to peptide tab file")
    parser.add_argument("--outputFile", required=True, help="Path to output file")
    parser.add_argument("--distinctPeptides", required=True, help="Path to distinct peptides file")
    parser.add_argument("--nonTaxaShortPeptideCutoff", required=True, help="Peptides less than this length from other taxa will not be output")
    return parser.parse_args()

def parseDistinctPeptides(file):
    rv = {}
    with open(file) as fh:
        for line in fh:
            peptide = line.strip()
            rv[peptide] = 1
    fh.close()
    return(rv)

def parsePeptideToIedb(file, keep):
    rv = {}
    with open(file) as fh:
        for line in fh:
            fields = line.strip().split("\t")
            peptide = fields[3]

            if peptide in keep:
                iedbId = fields[1]
                rv[peptide] = iedbId
    fh.close()
    return(rv)

def parseFullSeqMatchResults(file):
    rv = {}
    with open(file) as fh:
        for line in fh:
            peptide, protein, iedb, taxonMatch, proteinMatch = line.strip().split(",")

            if protein == "":
                protein = "NA"

            key = peptide + "_" + protein;

            vals = [taxonMatch, proteinMatch]
            rv[key] = vals
    fh.close()
    return(rv)

def main(args):

    gffSource = "veupathdb"
    gffType = "peptide"

    nonTaxaShortPeptideCutoff = int(args.nonTaxaShortPeptideCutoff)

    distinctPeptides = parseDistinctPeptides(args.distinctPeptides)
    peptideToIedb = parsePeptideToIedb(args.peptideTab, distinctPeptides)
    fullSeqMatchResults = parseFullSeqMatchResults(args.fullSeqMatchResults)

    removeSuffixPattern = r"\.\d+$"

    out = open(args.outputFile, 'w')
    with open(args.pepMatchResults) as fh:
        for line in fh:
            peptide, protein, mismatches, start, end = line.strip().split(",")

            # pepMatch preprocess db adds a ".1 suffix to proteins; remove it
            trimmedProtein = re.sub(removeSuffixPattern, "", protein)

            # this will throw KeyError exception if not found
            iedbId = peptideToIedb[peptide]

            key = peptide + "_" + trimmedProtein;
            altKey = peptide + "_NA"

            taxonMatch = 0
            fullProteinMatch = 0
            if key in fullSeqMatchResults:
                tm, pm = fullSeqMatchResults[key]
                taxonMatch = tm
                fullProteinMatch = pm
            elif altKey in fullSeqMatchResults:
                tm, pm = fullSeqMatchResults[altKey]
                taxonMatch = tm
                fullProteinMatch = pm
            else:
                taxonMatch = 0
                fullProteinMatch = 0

            gffId = trimmedProtein + "_" + start + "_" + end
            attributes = f"ID={gffId};iedb={iedbId};matchesTaxon={taxonMatch};matchesFullLengthProtein={fullProteinMatch};mismatches={mismatches};peptide={peptide}"

            gffRow = f"{trimmedProtein}\t{gffSource}\t{gffType}\t{start}\t{end}\t.\t.\t.\t{attributes}"

            if not (taxonMatch != 1 and len(peptide) < nonTaxaShortPeptideCutoff):
                print(gffRow, file=out)

    fh.close()

    out.close()

if __name__ == "__main__":
    main(parseArguments())

File: bin/mergeResults.pl
#!/usr/bin/env perl

use strict;

use Data::Dumper;
use Getopt::Long;


# This code macthes the outputs from the blast and the exact matches into a single text file. 
my ($pepMatchResults, $fullSeqMatchResults, $peptideTab, $distinctPeptidesFile, $outputFile);
&GetOptions("pepMatchResults=s"=> \$pepMatchResults,
            "fullSeqMatchResults=s" => \$fullSeqMatchResults,
            "peptideTab=s" => \$peptideTab,
            "outputFile=s"=> \$outputFile,
            "distinctPeptides=s" => \$distinctPeptidesFile,
    ) ;


my %distinctPeptides;
open(PEPTIDES, $distinctPeptidesFile) or die "Cannot open file $distinctPeptidesFile for reading: $!";
while(<PEPTIDES>) {
    chomp;
    $distinctPeptides{$_} = 1;
}
close PEPTIDES;

my %peptideToIedb;
open(PEPTAB, $peptideTab) or die "Cannot open file $peptideTab for reading: $!";
while(<PEPTAB>) {
    chomp;
    my @a = split(/\t/, $_);
    my $peptide = $a[3];
    next unless($distinctPeptides{$peptide});

    my $iedbId = $a[1];
    $peptideToIedb{$peptide} = $iedbId;
}
close PEPTAB;


my %fullSeqMatchResults;
open(FULLRES, $fullSeqMatchResults) or die "Cannot open file $fullSeqMatchResults for reading: $!";
while(<FULLRES>) {
    chomp;
    my ($peptide, $protein, $iedb, $taxonMatch, $proteinMatch) = split(",", $_);

    my $key = "${peptide}_${protein}";

    $fullSeqMatchResults{$key}->{taxon_match} = $taxonMatch;
    $fullSeqMatchResults{$key}->{full_protein_match} = $proteinMatch;

}
close FULLRES;

open(OUT, ">$outputFile") or die "Cannot open output file $outputFile for writing: $!";
open(PEPMATCH, $pepMatchResults) or die "Cannot open file $pepMatchResults for reading: $!";


while(<PEPMATCH>) {
    chomp;
    my ($peptide, $protein, $mismatches, $start, $end) = split(",", $_);

    # pepMatch preprocess db adds a ".1 suffix to proteins; remove it
    $protein =~ s/\.\d$//;

    my $iedbId = $peptideToIedb{$peptide};
    die "Could not map peptide to iedb id " unless $iedbId;

    my $key = "${peptide}_${protein}";

    my $taxonMatch = 0;
    my $fullProteinMatch = 0;

    if($fullSeqMatchResults{$key}) {
        $taxonMatch = 1 if $fullSeqMatchResults{$key}->{taxon_match};
        $fullProteinMatch = 1 if $fullSeqMatchResults{$key}->{full_protein_match};
    }

    print OUT join("\t", $peptide, $protein, $iedbId, $taxonMatch, $fullProteinMatch, $mismatches, $start, $end, "\n");
}


close PEPMATCH;
close OUT;


1;
